### ISSUE TITLE
Roadmap for official rectangular training and inference support?

### ISSUE THREAD

**Issue by qsq3** [roboflow/rf-detr#1052]
Hi RF-DETR team,

Is full rectangular training/inference officially supported, or still experimental? If it is still in progress, is there any ETA or recommended workaround for 16:9 datasets?

**Comment by isaacrob** [comment]
We don't have a specific plan to work on this for the time being. We recommend trying your dataset squashed to square and seeing how the accuracy works out. The model was pretrained to be able to handle distortions of that flavor.

**Comment by aholliday-kr** [comment]
I'd like to second this as an issue.  I began by squashing my dataset to square as @isaacrob recommended.  I then did a quick hacky change to make the code resize things to a 3:4 aspect ratio (the native ratio of my dataset) instead of to a square, trained a second model this way, and compared the results.  Using the native aspect ratio gave a considerable improvement in accuracy, around 20%.  So there are benefits to not having to distort non-square input images.  It would be great if I could do this without using my now somewhat out-of-date copy of the RF-DETR codebase with my hacky fix.

Incidentally, this might also be a reason to provide a DINOv3 backbone option, since as I understand it, DINOv3 models are more flexible about the input image shapes they will accept.

**Comment by isaacrob** [comment]
@aholliday-kr are you finetuning the pretrained model?

**Comment by aholliday-kr** [comment]
@isaacrob Yes, I am.

**Comment by isaacrob** [comment]
Got it. Could you link the change somewhere?

**Comment by aholliday-kr** [comment]
@isaacrob You can see my fork here: https://github.com/kinsolresearch/rf-detr
Note that I also added code to pad images that are the wrong aspect ratio instead of stretching them to the required ratio, since my dataset includes a minority of images that aren't 3:4.  This change was made later on, after the square-vs-3:4 experiment I described above.

**Comment by qsq3** [comment]
@aholliday-kr Thanks

**Comment by isaacrob** [comment]
cool! looked at your changes. so to be clear, you would want that at train time we allow you to specify an aspect ratio that gets enforced at train and test?

**Comment by aholliday-kr** [comment]
@isaacrob In a perfect world I wouldn't need to specify an aspect ratio at all: images would be handled at whatever their native asprect ratios were.  But specifying an aspect ratio at train time as you describe would certainly be an improvement.  It would also be nice to be able to control whether the specified ratio is to be enforced by stretching or by padding.

**Comment by isaacrob** [comment]
Got it. ViT models don't naturally handle variable aspect ratio, there's a lot of work on it and my favorite technique is NaViT but unfortunately that training technique isn't directly applicable to a DETR. Because of that, we kept our models square during pretraining. Other people have found significant decay when using non square aspect ratios, and while it seems to be working in your case I'm not really convinced that it'll work generally, especially without significant data. 

Having said that, it might be good to add the option. Just don't want to cause people to end up shooting themselves in the foot by accident!

Would you mind saying a bit more about your use case? How much data did you use, triple check that you used the pretrained weights, any other things worth noting about how you trained it that differ from just doing model.train?

**Comment by aholliday-kr** [comment]
Well, this is embarrassing.  Upon triple-checking, I realized that I had misunderstood a detail of the model-loading API and as a result had not been loading the pretrained weights at all this whole time.   I hadn't realized there was a problem because even without the pre-trained weights, my fine-tuned detector still had reasonable accuracy on my dataset.  I've now tried training with the pre-trained weights instead, which gives a decent improvement in accuracy.  I also reran my square-vs-3:4 experiment with the pre-trained weights, and found that there was no difference in accuracy between the two.  So @isaacrob is right after all: supporting rectangular training does not look worthwhile.

**Comment by anatoly-ryabchenko** [comment]
@isaacrob I actually think that the results - no difference in accuracy between square-vs-3:4 - is great! It means that during fine-tuning the model can adapt positional embedding / transformer blocks to work with rectangular images without loss in accuracy. There is still a great reason to allow rectangular input - latency!

RF-DETR has small models which are used in real-time constraints, and many real world cameras have 3:4 / 16:9 aspect ratio, so if you could have the same accuracy for the same resolution, but only use 55%-75% of the inference budget on that - that is a huge deal for these applications. Also it will allow faster fine-tuning (maybe not, since it may take longer time to adapt). Having worked a lot on edge devices - phones and Nvidia Jetson - I can say that it is really nice to have rectangular images support. 

If you agree, I can create a fix for it, evaluate on couple of datasets - to verify that accuracy doesn't change and also check how training times are affected - and create a PR.

**Comment by isaacrob** [comment]
@aholliday-kr when you did your square vs rectangular, did you control for token count? Or was the rectangular running with fewer tokens

**Comment by anatoly-ryabchenko** [comment]
> [@aholliday-kr](https://github.com/aholliday-kr) when you did your square vs rectangular, did you control for token count? Or was the rectangular running with fewer tokens

In worst case (rarely happens) - training with original image sizes - you have same width (or height) in square, but have to stretch in the other dimension: no additional information (similar to padding) in expense of compute. In best case - you downscale one or both dimensions, so one dimension preserves some more info wrt rectangle (or padding), but with rectangle you could just keep the aspect ratio and raise the resolution instead! In fact, NaViT paper that you mentioned found that for ViTs aspect-preserving beats resizing-to-square at matched token count, but it's also intuitive. And also the fact that you change the aspect ratio of the image implies the model needs to adopt to it during training as well - there's random jitter on aspect ratio during pre-training (up to 4:3), but that may not be enough for some cameras / effective image sizes (e.g. 16:9).

So rectangular training / inference seems to be naturally the best option, the real question is whether pre-trained RF-DETR can adapt to a rectangular ratio during fine-tuning on real world dataset. If you are interested, I can check this on some hard relatively small datasets with high variability / where aspect ratio is important (pedestrians) / small objects, similar to what the model might be trained on in real use cases - maybe VisDrone subset with 16:9 images?

As for the inference time, I checked and you have significant speedup from rectangular image wrt longer-side-equal-squares, especially on slower GPUs / with larger models where you are compute-bound: 

L4:
width / long side | Medium | 2XLarge
-- | -- | --
0.75 | 0.79 | 0.84
0.50 | 0.63 | 0.62
0.25 | 0.66 | 0.50

T4:
width / long side | Medium | 2XLarge
-- | -- | --
0.75 (0.75) | 0.74 | 0.74
0.50 (0.50) | 0.52 | 0.55
0.25 (0.25) | 0.35 | 0.36

I don't have a Jetson now unfortunately, but I expect even better, near-linear, improvements there.

**Comment by anatoly-ryabchenko** [comment]
Upon further reading, I realised I've had wrong assumptions initially - I assumed rf detr was letterboxing instead of squashing and didn't know that DINOv2 was already pretrained with jitter up to 4:3 in aspect ratio. So it wouldn't be fair to say we waste 25%-45% of budget. But in any case, native aspect ratio with same token count is supposed to be better (or at least not worse) for accuracy than square. So the question is whether the model can really adapt to new aspect ratio across datasets - @isaacrob have you evaluated that previously with your team by any chance? I am currently running an experiment on a single somewhat hard dataset, but I see that even with same token-count, the accuracy curve stays lower and plateaus as opposed to @aholliday-kr results - perhaps results can be on par with larger token-count (like keeping shorter side of rect equal to side of square).

**Comment by isaacrob** [comment]
We've heard through the grapevine lots of people trying to get it to work with non square and finding it works less well than squashing to square. I think it's unlikely to give strong results because the pretrain is square so the priors transfer less well to non squashed objects. Doing a pretrain with native mixed aspect ratio support is non trivial for DETR models with ViT backbones so we considered it out of scope but possibly relevant for a a v2 release.

**Comment by isaacrob** [comment]
I think unless your images are pathologically rectangular or you have very strange objects, the squash is probably fine. Though it may be more problematic for masks.

**Comment by LeMinhNgan** [comment]
> I think unless your images are pathologically rectangular or you have very strange objects, the squash is probably fine. Though it may be more problematic for masks.

@isaacrob,
Indeed we meet this type of image a lot in real application: the object is long-narrow shape, even 10:1 ratio,  and therefore we have to crop the original image to remove unnecessary and annoying context before bringing it to the object detection model. Then if we apply square shape we will face one of  two problems:
1. If we set normal resolution: one dimension of the image is resized to much to lost small objects
2. If we set high resolution: the GPU VRAM cannot adapt it.
Rectangular resolution is very useful and low cost approach. One of our images that i am testing is below. Hope you can add this function soon.

<img width="4033" height="514" alt="Image" src="https://github.com/user-attachments/assets/996aa091-cd37-4be7-913a-f51d84e88b42" />

**Comment by aholliday-kr** [comment]
@anatoly-ryabchenko Since you asked, in my rectangular experiment we kept the shorter image side length the same as in square and let the longer size be longer.  So the rectangular experiment uses more tokens per image than the square experiment.
